### PR TITLE
SONATA reports support

### PR DESCRIPTION
**Pull request by jorblancoa**
As seen in https://bbpteam.epfl.ch/project/issues/browse/BBPBGLIB-977, the results for LFP calculation using SONATA reports as input are different than using Binary reports when both contain the same values.

The issue was with the ordering of the GIDs in SONATA. The mapping created in EMSim was sorted by offset. In case of SONATA, the offset 0 wont necessary be the first GID.

The code was changed to sort the mapping first by cellIndex and second by sectionID. This way we ensure consistency in the vectors of positions and radius for both binary and SONATA reports.

### CODE REVIEW

**Comment by mgeplf** [comment]
Thanks! I won't have a chance to look into this today.  I'm surprised there is a difference, because at some point I did the same checking.  I will have to look at my notes.

**Comment by jorblancoa** [comment]
It was naive from my side to try and add a BlueConfig to the unit test... 
I cant think on other ways to thest the issue tho

**Comment by mgeplf** [comment]
> It was naive from my side to try and add a BlueConfig to the unit test... I cant think on other ways to thest the issue tho

That's fine; glad to have tests.

**Comment by mgeplf** [comment]
I think it would be worthwhile if @chevtche had a look, as he understands this better than I do.

**Comment by jorblancoa** [comment]
Since the CI don't have the infrastructure to access morphology paths (I guess we cant use BBPTestData in github), I will have to remove the unit test...

**Comment by chevtche** [comment]
Hi guys. The problems is just with a unit test ?

**Comment by jorblancoa** [comment]
Yes and no...
The unit test I would have to remove because there is no way to read a BlueConfig (since the morphology paths are not available and are needed by Brion).

However, could you have a look at the change itself?

**Comment by mgeplf** [comment]
@chevtche > The problems is just with a unit test ?

I think we need input, more, on the reason for sorting by offset here:
https://github.com/BlueBrain/EMSim/blob/master/emSim/EventsLoader.cpp#L243

IE: was this done for performance, or for correctness? Do the calculations have to be performed "emanating" out from from the somas?  From a perf standpoint, it seems weird to lose the locality of working w/ a morphology (ie: sorting per GID), so my suspicion is that there is another reason for this choice.

**Comment by chevtche** [comment]
Sorry for the late answer, I was pretty busy with the opendeck stuff.
I had a look at the code, but unfortunately I cannot really say if it's correct. I am not familiar with Sonata and this particular code was written by Juan Hernando. What I remember is that in the current report file the data was contiguous, all current values for all morphology segments for each simulation frame, frame after frame. You can see that Events->Powers has the same memory layout as the data in the current report file so one can just use memcpy() to update simulation frame. 

The goal of _computeStaticEventGeometry() is to find the corresponding segment or soma for each current data in the current report file. This is done once to place the events in 3d space (middle of the segment or soma position).

"Was this done for performance, or for correctness?"
As this is done only once I doubt it is for performance reasons...

In the end if you guys know the mapping between Sonata morphology segments end current reports just completely rewrite this function. Again it is only used to place the event (point current) at the correct 3d position. As morphologies are static this never changes during the simulation.

**Comment by jorblancoa** [comment]
>  You can see that Events->Powers has the same memory layout as the data in the current report file so one can just use memcpy() to update simulation frame.

That is the key, since events->powers() come from the data retrieved in loadFrame() (https://github.com/BlueBrain/EMSim/blob/master/emSim/EventsLoader.cpp#L50)
This loadFrame() returns the currents sorted by gid for SONATA reports as we can see in brion
https://github.com/BlueBrain/Brion/blob/master/brion/plugin/compartmentReportHDF5Sonata.cpp#L555
https://github.com/BlueBrain/Brion/blob/master/brion/plugin/compartmentReportHDF5Sonata.cpp#L396

However the radius and positions are added in the order defined by the mapping (remember that this was by offset).
https://github.com/BlueBrain/EMSim/blob/master/emSim/EventsLoader.cpp#L117

So basically at the end we were passing to the compute function the powers sorted by gid, while the rest of the vectors by offset (this will only work fine when the gids are sorted in the SONATA report).
https://github.com/BlueBrain/EMSim/blob/master/emSim/SamplePoints.cpp#L51

**Comment by chevtche** [comment]
I am not really familiar with the Brion code, not sure what this mapping is used for. But the currents cannot be sorted ONLY by GIDs as each morphology has a lot of segment. So there should be 2 level of sorting involved.

**Comment by mgeplf** [comment]
@jorblancoa tracked down the root cause:
```
My patch fixes the ordering of the mapping here https://github.com/BlueBrain/EMSim/blob/master/emSim/EventsLoader.cpp#L243 (from sorting by offset to sorting by cellIndex-sectionId pair).

In case of the ticket NSETM-1479, things work fine when the mapping is sorted by offset because the target of that simulation has more than 1000 cells. Why this matters? As you can see here https://github.com/BlueBrain/EMSim/blob/master/emSim/EventsLoader.cpp#L84 updateMapping() is called with a subset of the gids, since the morphologies are loaded in batches of 1000 as specified in line 70 of the same file.

This is telling brion that it is a subset and when trying to get the offsets from emSIM
https://github.com/BlueBrain/EMSim/blob/master/emSim/EventsLoader.cpp#L231 and
https://github.com/BlueBrain/Brion/blob/master/brion/plugin/compartmentReportHDF5Sonata.cpp#L238
 it will return the _targetMapping that is sorted by offset, while if your target is less than 1000 gids (like in this ticket’s example) will return the _sourceMapping that are in the order found in the h5 file.

To unify this 2 cases, and since the values from https://github.com/BlueBrain/EMSim/blob/master/emSim/EventsLoader.cpp#L50 will always be sorted by gid,
I think my patch is a simple fix to avoid having this inconsistency between the values and the mapping ordering.
```
Very nice work finding why we weren't always seeing differences.

**Comment by pramodk** [comment]
thank you @jorblancoa! (and @mgeplf and @chevtche as well!)

@mgeplf: to deploy this in production - do you plan to create a new tag?

**Comment by mgeplf** [comment]
@pramodk > to deploy this in production - do you plan to create a new tag?

I have tagged v1.0.2
